File: Methods-Java/Java source-code/chiSquare.java


public class chiSquare {
    public static double chiSquare_m(Double[] expected, Double[] observed) {
        double sumExpected = 0d;
        double sumObserved = 0d;
        for (int i = 0; i < observed.length; i++) {
            sumExpected += expected[i];
            sumObserved += observed[i];
        }
        double ratio = 1.0d;
        boolean rescale = false;
        if (Math.abs(sumExpected - sumObserved) > 10E-6) {
            ratio = sumObserved / sumExpected;
            rescale = true;
        }
        double sumSq = 0.0d;
        for (int i = 0; i < observed.length; i++) {
            if (rescale) {
                final double dev = observed[i] - ratio * expected[i];
                sumSq += dev * dev / (ratio * expected[i]);
            } else {
                final double dev = observed[i] - expected[i];
                sumSq += dev * dev / expected[i];
            }
        }
        return sumSq;
    }
}

File: Methods/Methods-Python/chiSquare.py
def chiSquare(expected, observed):
    sumExpected = 0
    sumObserved = 0

    for i in range(0, len(observed)):
        sumExpected += expected[i]
        sumObserved += observed[i]

    ratio = 1
    rescale = False

    if abs(sumExpected - sumObserved) > 10E-6:
        ratio = sumObserved / sumExpected
        rescale = True

    sumSq = 0

    for i in range(0, len(observed)):
        if rescale:
            dev = observed[i] - (ratio * expected[i])
            sumSq += (dev * dev) / (ratio * expected[i])
        else:
            dev = observed[i] - expected[i]
            sumSq += dev * dev / expected[i]

    return sumSq
